### PR TITLE
Added BinarySearch in C#

### LINKED ISSUE #373: BinarySearch in C#
This is an open issue for binary search algorithm implementation in C#. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by riturajjain2000**
**Fixes #373**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 }; 
+		int n = arr.Length; 
+		int x = 10; 
+		int result = DoBinarySearch(arr, x); 
+		if (result == -1) 
+			Console.WriteLine("Element not present"); 
+		else
+			Console.WriteLine("Element found at "
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:42]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	} 
+
+	//Driver program to test above function 
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 }; 
+		int n = arr.Length; 
+		int x = 10; 
+		int result = DoBinarySearch(arr, x); 
+		if (result == -1)
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:39]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			//1) Check if x is present at mid 
+			if (arr[m] == x) 
+				return m; 
+
+			//2) If x greater, ignore left half 
+			if (arr[m] < x) 
+				l = m + 1; 
+
+			//3) If x is smaller, ignore right half 
+			else
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:23]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'else' block.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		// if we reach here, then element was 
+		// not present 
+		return -1; 
+	} 
+
+	//Driver program to test above function 
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 }; 
+		int n = arr.Length;
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:36]: Codacy found an issue: [Remove the unused local variable 'n'.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	// else return -1 
+	static int DoBinarySearch(int[] arr, int x) 
+	{ 
+		int l = 0, r = arr.Length - 1; 
+		while (l <= r) {
+
+			int m = l + (r - l) / 2; //mid point
+
+			//1) Check if x is present at mid 
+			if (arr[m] == x)
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:15]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		while (l <= r) {
+
+			int m = l + (r - l) / 2; //mid point
+
+			//1) Check if x is present at mid 
+			if (arr[m] == x) 
+				return m; 
+
+			//2) If x greater, ignore left half 
+			if (arr[m] < x)
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:19]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	//Driver program to test above function 
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 }; 
+		int n = arr.Length; 
+		int x = 10; 
+		int result = DoBinarySearch(arr, x); 
+		if (result == -1) 
+			Console.WriteLine("Element not present");
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:40]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+// C# implementation of Binary Search 
+using System; 
+
+class BinarySearch {
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:4]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		int[] arr = { 2, 3, 4, 10, 40 };
+		int x = 10;
+		int result = DoBinarySearch(arr, x); 
+		if (result == -1) 
+		{
+			Console.Write("Element not present"); 
+		}
+		else
+		{
+			Console.Write("Element found at "
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:50]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+	//Driver program to test above function 
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 };
+		int x = 10;
+		int result = DoBinarySearch(arr, x); 
+		if (result == -1) 
+		{
+			Console.Write("Element not present");
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:46]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

Comment (riturajjain2000) [reply algorithms/ar-bsrh/csharp/BinarySearch.cs:46]: @RavinduSachintha Sir I am unable to resolve this issue , can you plz help me

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+		// if (result == -1) 
+		// {
+		// 	Console.Write("Element not present"); 
+		// }
+		// else
+		// {
+		// 	Console.Write("Element found at "
+		// 					+ "index " + result); 
+		// }
+		Console.Write(result);
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:53]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:53]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+		return -1; 
+	} 
+
+	//Driver program to test above function 
+	public static void Main() 
+	{ 
+		int[] arr = { 2, 3, 4, 10, 40 };
+		int x = 10;
+		int result = DoBinarySearch(arr, x); 
+		// if (result == -1)
```
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:44]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs (hunk 11 above)
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:53]: Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs (hunk 12 above)
Comment (RavinduSachintha) [inline algorithms/ar-bsrh/csharp/BinarySearch.cs:44]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6330839)

File: algorithms/ar-bsrh/csharp/BinarySearch.cs (hunk 11 above)
Comment (Danushka96) [reply algorithms/ar-bsrh/csharp/BinarySearch.cs]: @riturajjain2000 Remove or comment Line 53

File: algorithms/ar-bsrh/csharp/BinarySearch.cs (hunk 12 above)
Comment (Danushka96) [reply algorithms/ar-bsrh/csharp/BinarySearch.cs]: @riturajjain2000 Remove the commented line 44

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@riturajjain2000 good job :+1: keep it up.

**Comment by riturajjain2000** [comment]
> @riturajjain2000 good job 👍 keep it up.

Thank you sir 😄

**Pull request merged**